### PR TITLE
Use a single op for xattr get/set

### LINKED ISSUE #2483: Write cas to sync xattr
On Sync Gateway writes, use macro expansion to stamp the sync xattr with the document's updated cas value.

### PR DESCRIPTION
**Pull request by adamcfraser**
Uptakes https://issues.couchbase.com/browse/MB-23522.

Some additional handling required for GET to handle both the 'doc + xattr' and 'deleted doc + xattr' cases.

Also adds CAS stamping on write (closes #2483)

### CODE REVIEW

**Comment by adamcfraser** [comment]
Pending gocb changes being pushed to master (tested against fork).

File: base/bucket_gocb.go [hunk 1]
```diff
@@ -872,20 +873,24 @@ func (bucket CouchbaseBucketGoCB) WriteCasWithXattr(k string, xattrKey string, e
 	}
 
 	casOut = 0
-	// Otherwise, replace existing value
-	// TODO: As above, should be single operation
+	// Otherwise, replace existing valu
```
Comment (tleyden) [inline base/bucket_gocb.go]: Typo: `valu` -> `value`

File: base/bucket_gocb.go [hunk 2]
```diff
@@ -853,17 +853,18 @@ func (bucket CouchbaseBucketGoCB) WriteCas(k string, flags int, exp int, cas uin
...
-		newCas, err := bucket.Bucket.Insert(k, &v, uint32(exp))
-		if err != nil {
-			return 0, err
-		}
-		docFragment, err := bucket.Bucket.MutateIn(k, newCas, uint32(exp)).
-			UpsertEx(xattrKey, xv, gocb.SubdocFlagXattr|gocb.SubdocFlagCreatePath).
+		docFragment, err := bucket.Bucket.MutateInEx(k, gocb.SubdocDocFlagReplaceDoc, 0, uint32(exp)).
+			UpsertEx(xattrKey, xv, gocb.SubdocFlagXattr).
+			UpsertEx(xattrCasProperty, "${Mutation.CAS}", gocb.SubdocFlagXattr|gocb.SubdocFlagUseMacros).
+			UpsertEx("", v, gocb.SubdocFlagNone).
```
Comment (tleyden) [inline base/bucket_gocb.go]: Why is this needed?  `UpsertEx("", v, gocb.SubdocFlagNone).`  The `gocb.SubdocFlagNone` docs say "indicate no special behaviors", but in previous method call it's passing `gocb.SubdocFlagXattr|gocb.SubdocFlagUseMacros`.  Also what does it mean to pass `""` in path?

File: base/bucket_gocb.go [hunk 3]
```diff
@@ -853,17 +853,18 @@ func (bucket CouchbaseBucketGoCB) WriteCas(k string, flags int, exp int, cas uin
...
 func (bucket CouchbaseBucketGoCB) WriteCasWithXattr(k string, xattrKey string, exp int, cas uint64, v interface{}, xv interface{}) (casOut uint64, err error) {
 
+	// WriteCasWithXattr always stamps the xattr with the new cas using macro expansion, into a top-level property called 'cas'.
+	// This is the only use case for macro expansion today - if more cases turn up, should change the bucket API to handle this more generically.
+	xattrCasProperty := fmt.Sprintf("%s.cas", xattrKey)
+
 	// If cas=0, treat as insert
 	if cas == 0 {
-		// TODO: do this insert and update in a single op, when https://issues.couchbase.com/browse/MB-23522 is fixed.
 		// TODO: Once that's fixed, need to wrap w/ retry handling
```
Comment (tleyden) [inline base/bucket_gocb.go]: The old comment was:

> // TODO: do this insert and update in a single op, when https://issues.couchbase.com/browse/MB-23522 is fixed.

Can retry handling be added as part of this PR?

File: base/bucket_gocb.go [hunk 4]
```diff
@@ -907,41 +912,53 @@ func (bucket CouchbaseBucketGoCB) GetWithXattr(k string, xattrKey string, rv int
...
+			GetEx(xattrKey, gocb.SubdocFlagXattr).
+			GetEx("", gocb.SubdocFlagNone).
+			Execute()
 
-		// If key not found, continue to attempt to retrieve the xattr only.  Otherwise standard error handling.
-		if docErr != nil && docErr != gocbcore.ErrKeyNotFound {
-			shouldRetry = isRecoverableGoCBError(docErr)
-			return shouldRetry, docErr, cas
-		}
+		if err == nil {
```
Comment (tleyden) [inline base/bucket_gocb.go]: this is a pretty big if / else if / else .. what about changing to a `switch` and putting the last `else` in a `default` block?

File: db/crud.go [hunk 5]
```diff
@@ -46,6 +47,7 @@ func (db *DatabaseContext) GetDoc(docid string) (doc *document, err error) {
 		return nil, base.HTTPErrorf(400, "Invalid doc ID")
 	}
 	dbExpvars.Add("document_gets", 1)
+	log.Printf("Using Xattrs? %v", db.UseXattrs())
```
Comment (tleyden) [inline db/crud.go]: Can you change this to a `base.Log` or remove?

File: db/database.go [hunk 6]
```diff
@@ -1059,6 +1060,7 @@ func (context *DatabaseContext) GetUserViewsEnabled() bool {
 }
 
 func (context *DatabaseContext) UseXattrs() bool {
+	log.Printf("Use Xattrs: %v", context.Options.UnsupportedOptions.EnableXattr)
```
Comment (tleyden) [inline db/database.go]: Can you change this to a base.Log or remove?

**Review by tleyden — CHANGES_REQUESTED** [review]
See inline comments

File: db/database.go (hunk 6 above)
Comment (adamcfraser) [reply db/database.go]: Thanks for catching - should be removed.

File: base/bucket_gocb.go (hunk 2 above)
Comment (adamcfraser) [reply base/bucket_gocb.go]: UpsertEx("") is how we update the doc body along with the xattr - it's the full doc subdoc operation.

Comment (tleyden) [reply base/bucket_gocb.go]: Ok can you add a comment to the right of that line?

`UpsertEx("", v, gocb.SubdocFlagNone).  // update the doc body along with the xattr (full doc subdoc operation.)`

File: base/bucket_gocb.go (hunk 4 above)
Comment (adamcfraser) [reply base/bucket_gocb.go]: Updated.

File: base/bucket_gocb.go (hunk 3 above)
Comment (adamcfraser) [reply base/bucket_gocb.go]: Added.  Added a RetrySleeper builder to BucketSpec to reduce code duplication.
